### PR TITLE
Add simple curve cycle trait for curve cycles

### PR DESCRIPTION
**Pull request by drewstone**
## Description
The purpose of this PR is to define a trait for standard (potentially non-pairing based) and pairing based curve cycles.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`

No unit tests as this is just new trait def.

### CODE REVIEW

**Comment by drewstone** [comment]
Ngl I just wanna contribute to this awesome project. I am still wearing my training wheels when looking around. Let me know if this isn't the right place or implementation for something like this. I have yet to see a PCD or cycle engine defined for non-pairing based curves so I figured there ought to be one.

**Comment by daira** [comment]
Looks good. Can you impl this for Pallas and Vesta?

**Comment by Pratyush** [comment]
This makes me think we need to have a `CurveCycle` trait, like so:

```rust
pub trait CurveCycle 
where
    <Self::E1 as AffineCurve>::Projective: MulAssign<<Self::E2 as AffineCurve>::BaseField>,
    <Self::E2 as AffineCurve>::Projective: MulAssign<<Self::E1 as AffineCurve>::BaseField>,
{
	// Maybe these constraints are too tight and could cause rust's constraint solver
	// error out due to cycles (heh); we remove the
	// bounds on the associated types if that's the case.
	type E1: AffineCurve<
		BaseField = <Self::E2 as AffineCurve>::ScalarField,
		ScalarField = <Self::E2 as AffineCurve>::BaseField,
	>;
	type E2: AffineCurve<
		BaseField = <Self::E1 as AffineCurve>::ScalarField,
		ScalarField = <Self::E1 as AffineCurve>::BaseField,
	>;
}

pub trait PairingFriendlyCycle: CurveCycle {
	// add whatever extra bounds are necessary.
	type Engine1: PairingEngine<
		G1Affine = Self::E1,
		G1Projective = <Self::E1 as AffineCurve>::Projective,
		Fq = <Self::E1 as AffineCurve>::BaseField,
		Fr = <Self::E1 as AffineCurve>::ScalarField,
	>;
	
	type Engine2: PairingEngine<
		G2Affine = Self::E2,
		G2Projective = <Self::E2 as AffineCurve>::Projective,
		Fq = <Self::E2 as AffineCurve>::BaseField,
		Fr = <Self::E2 as AffineCurve>::ScalarField,
	>;
}
```

**Comment by drewstone** [comment]
@daira where's the right place to implement that?

**Comment by drewstone** [comment]
I've updated the trait with your feedback @Pratyush. Getting rid of the strict def on `E2` managed to get it to compile without issues. Let me know your thoughts!

File: ec/src/lib.rs [hunk 1]
```diff
@@ -313,6 +313,8 @@ pub fn prepare_g2<E: PairingEngine>(g: impl Into<E::G2Affine>) -> E::G2Prepared
 }
 
 /// A cycle of pairing-friendly elliptic curves.
+/// TODO: remove this once dependencies are
+///       updated with new curve cycle traits
```
Comment (Pratyush) [inline ec/src/lib.rs:317]: Maybe we can add a `#[deprecated("Please use `PairingFriendlyCycle` instead")]` marker here?

File: ec/src/lib.rs [hunk 2]
```diff
@@ -313,6 +313,7 @@ pub fn prepare_g2<E: PairingEngine>(g: impl Into<E::G2Affine>) -> E::G2Prepared
 }
 
 /// A cycle of pairing-friendly elliptic curves.
+#[deprecated("Please use `PairingFriendlyCycle` instead")]
```
Comment (Pratyush) [inline ec/src/lib.rs:316]: oops, this should be

```suggestion
#[deprecated(note = "Please use `PairingFriendlyCycle` instead")]
```

**Comment by Pratyush** [comment]
Thanks for the PR! could you add a relevant entry in CHANGELOG.md under `Improvements`? Oh and a final format.

**Comment by drewstone** [comment]
All set @Pratyush

**Review by Pratyush — APPROVED** [review]

**Pull request merged**